libbpf-tools/drsnoop.c

```diff
@@ -0,0 +1,257 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Wenbo Zhang
+//
+// Based on drsnoop(8) from BCC by Wenbo Zhang.
+// 28-Feb-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/resource.h>
+#include <unistd.h>
+#include <time.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "drsnoop.h"
+#include "drsnoop.skel.h"
+#include "trace_helpers.h"
+
+static struct env {
+	pid_t pid;
+	pid_t tid;
+	time_t duration;
+	int page_cnt;
+	int timeout_ms;
+	bool extended;
+	bool verbose;
+} env = {
+	.page_cnt = 64,
+	.timeout_ms = 100,
+};
+
+const char *argp_program_version = "drsnoop 0.1";
+const char *argp_program_bug_address = "<ethercflow@gmail.com>";
+const char argp_program_doc[] =
+"Trace direct reclaim latency.\n"
+"\n"
+"USAGE: drsnoop [--help] [-p PID] [-t TID] [-d DURATION] [-e] [-P PAGECNT] "
+" [-T TIMEOUT]\n"
+"\n"
+"EXAMPLES:\n"
+"    drsnoop         # trace all direct reclaim events\n"
+"    drsnoop -p 123  # trace pid 123\n"
+"    drsnoop -t 123  # trace tid 123 (use for threads only)\n"
+"    drsnoop -d 10   # trace for 10 seconds only\n"
+"    drsnoop -e      # trace all direct reclaim events with extended faileds\n";
+
+static const struct argp_option opts[] = {
+	{ "duration", 'd', "DURATION", 0, "Total duration of trace in seconds" },
+	{ "extended", 'e', NULL, 0, "Extended fields output" },
+	{ "pid", 'p', "PID", 0, "Process PID to trace" },
+	{ "pagecnt", 'P', "PAGECNT", 0, "Perf buffer size (default 64)" },
+	{ "tid", 't', "TID", 0, "Thread TID to trace" },
+	{ "timeout", 'T', "TIMEOUT", 0, "Timeout of perf buffer poll in ms"
+					" (default 100)" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static int page_size;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int pid, page_cnt, timeout_ms;
+	time_t duration;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		errno = 0;
+		duration = strtol(arg, NULL, 10);
+		if (errno || duration <= 0) {
+			fprintf(stderr, "invalid DURATION: %s\n", arg);
+			argp_usage(state);
+		}
+		env.duration = duration;
+		break;
+	case 'e':
+		env.extended = true;
+		break;
+	case 'p':
+		errno = 0;
+		pid = strtol(arg, NULL, 10);
+		if (errno || pid <= 0) {
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		env.pid = pid;
+		break;
+	case 't':
+		errno = 0;
+		pid = strtol(arg, NULL, 10);
+		if (errno || pid <= 0) {
+			fprintf(stderr, "invalid TID: %s\n", arg);
+			argp_usage(state);
+		}
+		env.tid = pid;
+		break;
+	case 'P':
+		errno = 0;
+		page_cnt = strtol(arg, NULL, 10);
```

Comment:
check it's power of two:
```
if (page_cnt & (page_cnt - 1))
    fprintf(stderr, "PAGECNT has to be power of 2\n");
```

Alternatively (and probably a bit more user friendly), would be to just find closest power of two number